File: _designs_001/Read N Characters Given Read4/ReaderN.java
public class ReaderN {
    private Reader4 reader4;

    ReaderN(Reader4 reader4) {
        this.reader4 = reader4;
    }

    public int read(char[] buf, int n) {
        char[] buffer = new char[4];
        int readBytes = 0;
        boolean eof = false;
        while (!eof && readBytes < n) {
            int sz = reader4.read(buffer);
            if (sz < 4)
                eof = true;
            int bytes = Math.min(n - readBytes, sz);
            System.arraycopy(buffer, 0, buf, readBytes, bytes);
            readBytes += bytes;
        }
        return readBytes;
    }
}

File: _designs_001/Find the Celebrity/ReaderN.rb
class ReaderN
  def initialize(reader4)
    @reader4 = reader4
  end

  def read(buf, n)
    box4, total_bytes, eof = '', 0, false
    while (!eof and total_bytes < n)
      box_size = @reader4.read(box4)
      eof = true if box_size < 4
      read_size = [n - total_bytes, box_size].min
      buf << box4[0...read_size]
      total_bytes += read_size
    end
    total_bytes
  end
end
